File: Majority Element II/Majority_Element_II.java
class Solution {
    public List<Integer> majorityElement(int[] nums) {
        int first = 0, second = 0, cnt1 = 0, cnt2 = 0;
        for (int i : nums) {
            if (i == first) {
                cnt1++;
            } else if (i == second) {
                cnt2++;
            } else if (cnt1 == 0) {
                first = i;
                cnt1 = 1;
            } else if (cnt2 == 0) {
                second = i;
                cnt2 = 1;
            } else {
                cnt1--;
                cnt2--;
            }
        }
        cnt1 = 0;
        cnt2 = 0;
        for (int i : nums) {
            if (i == first) cnt1++;
            else if (i == second) cnt2++;
        }
        int n = nums.length;
        List<Integer> ans = new ArrayList<>();
        if (cnt1 > n / 3) ans.add(first);
        if (cnt2 > n / 3) ans.add(second);
        return ans;
    }
}

File: Majority Element II/Majority_Element_II.py
class Solution:
    def majorityElement(self, nums: List[int]) -> List[int]:
        first, second, cnt1, cnt2 = 0, 0, 0, 0
        for i in nums:
            if i == first:
                cnt1 += 1
            elif i == second:
                cnt2 += 1
            elif cnt1 == 0:
                first = i
                cnt1 = 1
            elif cnt2 == 0:
                second = i
                cnt2 = 1
            else:
                cnt1 -= 1
                cnt2 -= 1
        
        cnt1, cnt2 = 0, 0
        for i in nums:
            if i == first:
                cnt1 += 1
            elif i == second:
                cnt2 += 1
        
        ans = []
        n = len(nums)
        if cnt1 > n // 3:
            ans.append(first)
        if cnt2 > n // 3:
            ans.append(second)
        
        return ans
        